adapter-base/src/main/java/org/eclipse/hono/adapter/auth/device/jwt/JwtAuthProvider.java

```diff
@@ -112,9 +115,18 @@ protected Future<DeviceUser> doValidateCredentials(final JwtCredentials deviceCr
         final Promise<DeviceUser> result = Promise.promise();
         currentContext.executeBlocking(blockingCodeHandler -> {
             log.debug("validating JWT on vert.x worker thread [{}]", Thread.currentThread().getName());
-            if (checkJwtValidity(deviceCredentials, credentialsOnRecord)) {
+            final Jws<Claims> jws = getJws(deviceCredentials, credentialsOnRecord);
+            if (jws != null) {
                 blockingCodeHandler
-                        .complete(new DeviceUser(deviceCredentials.getTenantId(), credentialsOnRecord.getDeviceId()));
+                        .complete(new DeviceUser(deviceCredentials.getTenantId(), credentialsOnRecord.getDeviceId()) {
+
+                            private final Instant expirationTime = jws.getBody().getExpiration().toInstant();
+
+                            @Override
+                            public boolean expired() {
+                                return expirationTime.plusSeconds(ExternalJwtAuthTokenValidator.ALLOWED_CLOCK_SKEW).isBefore(Instant.now());
```

Comment:
instead of doing this calculation here, referring to the protected  *ExternalJwtAuthTokenValidator.ALLOWED_CLOCK_SKEW* field, we could simply add a static *Instant ExternalJwtAuthTokenValidator.getExpirationTime(Instant)* which computes the expiration time based on the allowed clock skew. FMPOV this would be preferable because it doesn't decrease encapsulation ...

Reply:
Sounds good, implemented like suggested.